### PR TITLE
Ss brake emergency

### PR DESCRIPTION
**Pull request by Knhwong**
First commit

### CODE REVIEW

File: lib/brakes/brakes.cpp [hunk 1]
```diff
@@ -0,0 +1,17 @@
+#include "brakes.hpp"
+
+brakes::BrakeClass(const std::uint8_t pin_, const hyped::io::HardwareGpio gpio)
```
Comment (maxguy2001) [inline lib/brakes/brakes.cpp:3]: Naming - `BrakeClass`?

File: lib/brakes/brakes.cpp [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+{
+  pin    = pin_;
```
Comment (maxguy2001) [inline lib/brakes/brakes.cpp:5]: Think you're getting naming conventions mixed up here. Class data members have an underscore at the end, function arguments do not (i.e. should be `pin_=pin`).

File: lib/brakes/brakes.hpp [hunk 3]
```diff
@@ -0,0 +1,18 @@
+#pragma once
+#include <cstdint>
+#include <memory>
+#include <optional>
```
Comment (maxguy2001) [inline lib/brakes/brakes.hpp:4]: not used

File: lib/brakes/brakes.hpp [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+#include <core/types.hpp>
+#include <io/gpio.hpp>
+
+class BrakeClass {
+ public:
+  BrakeClass(const std::uint8_t pin_, const hyped::io::HardwareGpio gpio);
```
Comment (maxguy2001) [inline lib/brakes/brakes.hpp:11]: documention.

File: lib/io/hardware_gpio.cpp [hunk 5]
```diff
@@ -1,36 +1,125 @@
...
 namespace hyped::io {
 
 std::optional<core::DigitalSignal> HardwareGpioReader::read()
 {
-  // TODOLater: implement
-  throw -1;
+  // pinMap just 0000.... with 1 flipped in pin number.
+  // So AND with readAddr to extract the specific pin
+  const std::uint8_t out = 0;
+  if (*gpio_readAddr & pin_map) {
```
[the remark this answers is not in the document]
Comment (maxguy2001) [reply lib/io/hardware_gpio.cpp:15]: and `pin_map_`

File: lib/io/hardware_gpio.cpp [hunk 6]
```diff
@@ -1,36 +1,125 @@
...
+  // What happens if all shared pointers are removed and this is called?
+  // Map solves this, GPIO will always hold one pointer so it dosen't get destroyed.
+  if (initialized_readers_.count(pin) != 0) {
+    std::shared_ptr<IGpioReader> reader = initialized_readers_[pin];
+    return reader;
+  }
+
+  // There are 4 Bank Numbers, 0 1 2 3
+  // Integer divison to get bank number
+  const std::uint32_t bank = pin / 32;
```
[the remark this answers is not in the document]
Comment (maxguy2001) [reply lib/io/hardware_gpio.cpp:48]: and elsewhere. Don't think you need uint32_t here?

File: lib/io/hardware_gpio.cpp [hunk 7]
```diff
@@ -1,36 +1,125 @@
...
+  }
+  // Type conversion for address adding from void
+  const std::uint64_t base = reinterpret_cast<std::uint64_t>(gpio_addr);
+  // pinRead is the hardware specified address for reading.
+  gpio_read = reinterpret_cast<volatile std::uint32_t *>(base + pin_read);
+
+  // Keep track of intialized pins
+  initialized_readers_[pin]
+    = std::shared_ptr<HardwareGpioReader>(new HardwareGpioReader(pin_map, gpio_read));
+  std::shared_ptr<IGpioReader> reader = initialized_readers_[pin];
```
[the remark this answers is not in the document]
Comment (maxguy2001) [reply lib/io/hardware_gpio.cpp:84]: For sake of consistency in the codebase.

File: lib/io/hardware_gpio.hpp [hunk 8]
```diff
@@ -2,31 +2,63 @@
...
 #include "gpio.hpp"
 
+#include <unordered_map>
+
 #include <core/types.hpp>
 
 namespace hyped::io {
 
+// GPIO hardware specified addresses and sizes for read, clear, set, size, etc.
+static constexpr std::uint32_t pin_size  = 0x1000;
```
Comment (maxguy2001) [inline lib/io/hardware_gpio.hpp:12]: `kPinSize`?

Comment (maxguy2001) [reply lib/io/hardware_gpio.hpp:12]: Also this feels like quite a high scope to be declaring these? Is there absolutely nothing else in `hyped::io` that could be called `pin_size`?

File: lib/io/hardware_gpio.hpp [hunk 9]
```diff
@@ -2,31 +2,63 @@
...
  private:
-  HardwareGpioWriter(const std::uint8_t pin);
+  HardwareGpioWriter(const std::uint8_t pin,
+                     volatile std::uint32_t *set,
+                     volatile std::uint32_t *clear)
+      : pin_map(pin),
+        gpio_setAddr(set),
+        gpio_clearAddr(clear){};
+
+  const std::uint8_t pin_map;
```
Comment (maxguy2001) [inline lib/io/hardware_gpio.hpp:51]: naming

**Review by maxguy2001 — CHANGES_REQUESTED** [review]

**Comment by Knhwong** [comment]
Changed to BrakesClean to base of main.

**Pull request closed without merging**